src/java/main/org/apache/zookeeper/server/DataTree.java

```diff
@@ -732,6 +736,64 @@ public Stat statNode(String path, Watcher watcher)
         }
     }
 
+    /**
+     * This will return a paginated list of children for a given path
+     * @param path          The path we want to list
+     * @param stat          The current stat for this path
+     * @param watcher       The watcher to attach to this path. The watcher will be added only when we return <= maxReturned items
+     * @param maxReturned   How many to return max. We will return one more than this number to indicate truncation
+     * @param minZkId       The last zkID to not return. Any returned item will have a zkid > minZkId
+     * @return              A list of path with stats
+     * @throws NoNodeException
+     */
+    public List<PathWithStat> getPaginatedChildren(String path, Stat stat, Watcher watcher, int maxReturned, long minZkId) throws NoNodeException {
+        DataNode n = nodes.get(path);
+        if (n == null) {
+            throw new KeeperException.NoNodeException();
+        }
+        synchronized (n) {
+            if (stat != null) {
+                n.copyStat(stat);
+            }
+            PriorityQueue<PathWithStat> children;
+            Set<String> childs = n.getChildren();
+            if (childs == null) {
+                children = new PriorityQueue<PathWithStat>(1);
+            } else {
+                children = new PriorityQueue<PathWithStat>(maxReturned + 1, new Comparator<PathWithStat>() {
+                    @Override
+                    public int compare(PathWithStat o1, PathWithStat o2) {
+                        final long l = o2.getStat().getMzxid() - o1.getStat().getMzxid();
+                        return l < 0 ? -1 : (l == 0 ? 0 : 1);
+                    }
+                });
+                for (String child: childs) {
+                    DataNode childNode = nodes.get(path + "/" + child);
+                    if (null != childNode) {
+                        final long mzxid = childNode.stat.getMzxid();
```

Comment:
Currently the comparison logic uses mzxid. There could be chance of returning the same child node again. Assume the child node has been modified after returning the first batch. Now mxid will be the updated with the latest zxid which results in returning the same node again in the next batch query, right?. Please use Stat#getCzxid(). czxid - The zxid of the change that caused this znode to be created.